AnnounceSignatures resend in gossiper.

halseth opened the pull request:
This PR adds state of AnnouceSignatures to the gossiper, making it retry sending it to the remote peer if it fails initially. This state is also persisted on restarts, making it retry sending the message if need be.

This should help the case where a peer has gone offline before receiving the announcement signatures, not making us able to craft the final channel proof, even though the peer comes back online.

Roasbeef commented on discovery/gossiper.go:
```diff
@@ -1221,9 +1275,25 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
-				"for the channel with chanID=%v: %v", msg.ChannelID, err)
+		// Now set the waitingproof completed, such that it can be
+		// removed from the database if it was sent to the peer.
+		err = d.waitingProofs.SetCompleted(msg.ShortChannelID, true)
```
It seems like there's a possible race between this, and the goroutine launched below. This has the potential to cause the waiting proofs to stick around on disk longer than is strictly required. If we crash _before_ we get to this point, then we never mark the proof as completed. As a result, it will never be deleted from disk. 

A remedy to this would be to scan all waiting proofs on startup and delete the iff, if we read the channel edge from disk, it's announcement is already set. I think this works?

cfromknecht commented on channeldb/waitingproof.go:
```diff
@@ -3,106 +3,261 @@ package channeldb
+
+	// Completed indicated if the proof was fully assembled and added
+	// to the router graph, ready to be announced to the network.
+	Completed bool
```
Do we ever read this variable? AFAICT it only ever gets assigned

halseth commented on channeldb/waitingproof.go:
```diff
@@ -3,106 +3,261 @@ package channeldb
+		}
+
+		// Put the peer into this channel bucket.
+		return putPeer(channelBucket, remotePeer)
```
yeah, _almost_ enough. by fetching the channel info for that particular channel ID we have the public keys of the two peers of the channel, but I couldn't figure out a clever way of figuring out _which_ of the two is the remote, and which is the local node. 

If there's an elegant way of fetching this missing information, this could be simplified, and we wouldn't need to store the peer id in the database.

halseth commented on channeldb/waitingproof.go:
```diff
@@ -3,106 +3,261 @@ package channeldb
-	if _, ok := s.cache[key]; !ok {
-		return ErrWaitingProofNotFound
-	}
+// SetSent marks this proof's value indicating if it is sent to the remote peer
```
We need to keep the proof around until we have received the remote part as well.

halseth commented on discovery/gossiper.go:
```diff
@@ -1272,6 +1324,63 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
+
+	// As the message was successfully sent to the remote peer, mark the
+	// proof as sent in the database.
+	if err := d.waitingProofs.SetSent(msg.ShortChannelID, true); err != nil {
```
This was written with the underlying assumption that the when `SendToPeer` returns a success, then the message is delivered to the peer. Not sure how we can actually be sure it is delivered, and introducing an ACK would deviate from the protocol? I think delivery+ack is the responsibility of the transport in this case, and we should make changes there (server+peer+brontide) to make sure it is actually delivered.

halseth commented:
I ended up rewriting this quite a bit, reverting the changes done to the `WaitingProofStore`, and instead storing the result of sending the announcement message in a separate database within the gossiper. This lead to a much smaller diff, and simpler logic, since the two set of states (`Sent` and `Completed`) now is handled separately. This should also get rid of the mentioned race. 

In addition I added the suggested changes to the peer and server (see the fixup! commits), making the the queueing of messages not blocking, and added to the godoc an explanation of how to use the method to broadcast messages.

cfromknecht commented on discovery/gossiper.go:
```diff
@@ -1347,6 +1397,106 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
+
+		// Since the message was successfully sent to the peer, we
+		// can safely delete it from the database.
+		err := d.cfg.DB.Update(func(tx *bolt.Tx) error {
```
Even with the above changes, I believe there might still be an issue if the message is sent successfully, the remote peer crashes, and then comes back online as we won't attempt to resend the message. Restarting lnd would fix this, but obviously not ideal.

The simplest solution that comes to mind is to not break the loop after a successful send, but to sleep for some amount of time and then check the on-disk waiting proof store. If the entry is still in the waiting proof store, we can `continue` and just resend as if it was the first attempt, only breaking if it has been removed due to having received the full announcement as in my comment above.

The behavior of this loop would then be: periodically send while peer is online, blocking using server's notifications during intermittent disconnects, until the waiting proof entry has been removed.

Roasbeef replied:
See my comment in the start method, that should take care of this case. With that we'll: check all our outgoing channels, and re-send the ann sig (using this func) for those that we don't have a full chan auth proof for (and isn't private). This simple bit of logic meets our goals: we'll keep resending the ann-sig until we either get one back, or the other side broadcasts the full chan ann message.

cfromknecht commented on peer.go:
```diff
@@ -1122,12 +1122,20 @@ func (p *peer) PingTime() int64 {
-		return
+		peerLog.Debugf("Peer shutting down, could not enqueue msg.")
+		err := fmt.Errorf("peer shutting down")
+		if errChan != nil {
```
If we're returning the error directly, do we need to add this to the channel? Since the error indicates that message wasn't queued at all, a non-nil would signify that reading from the `errChan` isn't necessary.

Roasbeef replied:
Yeh I don't think we need to return the error. Instead we can just send over the channel as that lets the caller process the error in an async manner. Either is needed, not both though IMO.

cfromknecht commented on server.go:
```diff
@@ -982,9 +984,32 @@ func (s *server) sendPeerMessages(
+		errChans = append(errChans, errChan)
 	}
+
+	// Return an error if any of the messages failed being sent.
```
Nice job here as well! This has much better interleaving properties :) 

For both sends and broadcasts, this will still hold the server's mutex until all messages have been ACKd or failures occurs. With a few small changes, we can make this only block during queueing, and allow the receiver to conditionally handle errors outside the mutex. This should help keep large sends or broadcasts from limiting our throughput and minimize contention in the server.

1. Have `sendPeerMessages` and `sendToPeer` return `([]chan error, error)`
2. Remove the remainder of this section into a separate method such as `s.handleErrChans([]chan error) error`
3. Modify `SendToPeer` to handle error chans, such as:
```golang
s.Lock()
errChans, err := s.sendToPeer(target, msgs)
s.Unlock()
if err != nil {
    return err
}

return s.handleErrChans(errChans)
```

We won't need to modify broadcasting at all because we don't care about the errors and this actually returns it to the desired send-and-forget behavior. I suspect the compiler might even be able to optimize out the process of appending errChans during broadcasts, since the return variable can't be captured when run as a go routine.

cfromknecht replied:
If the compiler optimization doesn't happen, we can always shave the extra memory allocations by using the nil-ness of the wait group to instruct whether or not to instantiate error chans and appending to the slice.

halseth replied:
Good idea. If doing it this way, I would argue that we might just make the methods return _only_ `[]chan error`. Since a broadcast does not distinguish between failing to add a message to the queue, and failure sending at a later point, and a "reliable send" will also treat these two error scenarios similarly, we might as well only check the error channels if we want to.

The only downside I see by doing this is that if we send a batch of messages, and the first one cannot be queued, we will continue trying to queue the rest regardless. I don't think this will usually be a problem (they will all just fail quickly), but in the unlikely scenario that one succeeds (while a previous one failed) we end up with messages out of order. Not even sure this is possible 😛 

We _can_ make `sendPeerMessages` return only `[]chan error` as mentioned above, but still keep the return error from `queueMsg`. In the case of failing to queue a message, we can send this error directly on the remaining error chans, avoiding the problem I mentioned above.

Thoughts?

Roasbeef commented on discovery/gossiper_test.go:
```diff
@@ -847,9 +1236,319 @@ func TestOrphanSignatureAnnouncement(t *testing.T) {
+
+// Test that we retry sending the AnnounceSignatures to the remote peer,
+// even though we have been able to receive both parts of the proof.
+func TestSignatureAnnouncementRetryAfterBroadcast(t *testing.T) {
```
Doesn't seem like this is needed. If we have the full proof (a fully signed chan ann message), then we don't need to send it to the remote peer, as upon reconnect, we'll just send them the full proof (which is what we're trying to create anyway).

Roasbeef commented on peer.go:
```diff
@@ -1122,12 +1122,20 @@ func (p *peer) PingTime() int64 {
 // queueMsg queues a new lnwire.Message to be eventually sent out on the
-// wire.
-func (p *peer) queueMsg(msg lnwire.Message, doneChan chan struct{}) {
+// wire. It returns an error if we failed queing the message. An error
```
queing -> failed to queue

Roasbeef commented on server.go:
```diff
@@ -982,9 +984,32 @@ func (s *server) sendPeerMessages(
 	for _, msg := range msgs {
-		targetPeer.queueMsg(msg, nil)
+		errChan := make(chan error, 1)
+		if err := targetPeer.queueMsg(msg, errChan); err != nil {
```
If we modify `queueMsg` to just send an error, or return one (not both), then we don't also need to check the error here. As otherwise, it may _always_ block.

Roasbeef commented on server.go:
```diff
@@ -982,9 +984,32 @@ func (s *server) sendPeerMessages(
+	// Return an error if any of the messages failed being sent.
+	for _, errChan := range errChans {
+		select {
+		case err := <-errChan:
```
As was pointed out before, we don't _always_ want this method to block until completion. Instead, it can take the  error channel, and only block if it's nil.

halseth commented:
Another day, another revamp! Notable changes this time is that the resending logic is significantly simplified:
1. At startup we check all local proofs in the message store, and send them if we don't have the full proof. If we do, we just delete the local proof from the message store.
2. When sending the local proof, we add it to the message store first, but we don't delete it after a successful send. This because we cannot really know if the remote was able to store it. We will delete it at first gossiper startup after we have the full proof stored.
3. If we see the remote sending us his remote proof while we already have assembled the full proof, we send the full proof to the remote.
4. To make (3) possible, an additional check is added when attempting to add a `ChannelAnnouncement` to the `router`: if `AddEdge` fails with `ErrIgnored/ErrOutdated` we check if the edge we have in the graph already contains the channel proof, if not we add it at this point.

Roasbeef commented on discovery/gossiper.go:
```diff
@@ -319,6 +335,14 @@ func (d *AuthenticatedGossiper) Start() error {
 	d.bestHeight = height
 
+	// In case we had an AnnounceSignatures ready to be sent when the
+	// gossiper was last shut down, we must continue on our quest to
```
Onwards! 🤺

Roasbeef commented on discovery/gossiper.go:
```diff
@@ -518,6 +542,145 @@ func (d *deDupedAnnouncements) Emit() []lnwire.Message {
+		if chanInfo.AuthProof != nil {
+			log.Debugf("Deleting message for chanID=%v from "+
+				"messageStore", t.msg.ChannelID)
+			if err := d.cfg.DB.Update(func(tx *bolt.Tx) error {
```
Seems like `deleteMsg` can be re-used here?

halseth replied:
Yep, that was the reason I created the method in the first place 🤦‍♂️

Roasbeef commented on discovery/gossiper.go:
```diff
@@ -991,8 +1154,74 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
 			if routing.IsError(err, routing.ErrOutdated,
 				routing.ErrIgnored) {
 
+				// The edge will get rejected if we already
```
Nice! 

However, the indentation is a bit much. Would recommend rolling this into it's own function. This isn't a blocker on the PR though.

halseth replied:
Added TODO

Roasbeef commented on discovery/gossiper.go:
```diff
@@ -1190,6 +1420,73 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
+
+		// Check if we already have the full proof for this channel.
+		if chanInfo.AuthProof != nil {
+
```
Nit: extra space here.
